### PR TITLE
Improved Resource ownership

### LINKED ISSUE #35: [BUG] TestNext.TapUniquePtr intermittently fails
**Describe the bug**
```
[ RUN      ] TestNext.TapUniquePtr
terminate called after throwing an instance of 'std::system_error'
  what():  Resource deadlock avoided
*** Aborted at 1655311872 (unix time) try "date -d @1655311872" if you are using GNU date ***
PC: @                0x0 (unknown)
*** SIGABRT (@0x432) received by PID 1074 (TID 0x7fd64d35e000) from PID 1074; stack trace: ***
    @     0x7fd65a2f5197 google::(anonymous namespace)::FailureSignalHandler()
    @     0x7fd65a288420 (unknown)
    @     0x7fd659e9e00b gsignal
    @     0x7fd659e7d859 abort
    @     0x7fd65a118036 __gnu_cxx::__verbose_terminate_handler()
    @     0x7fd65a116524 __cxxabiv1::__terminate()
    @     0x7fd65a1158c5 __cxa_call_terminate
    @     0x7fd65a115e6f __gxx_personality_v0
    @     0x7fd65a05ed79 _Unwind_RaiseException_Phase2
    @     0x7fd65a05f07c _Unwind_RaiseException
    @     0x7fd65a11675b __cxa_throw
    @     0x7fd65a113479 std::__throw_system_error()
    @     0x7fd65a133969 std::thread::join()
    @     0x7fd65ba24c0e srf::internal::system::FiberTaskQueue::~FiberTaskQueue()
    @     0x7fd65ba228b9 __gnu_cxx::new_allocator<>::destroy<>()
    @     0x7fd65ba22799 std::allocator_traits<>::destroy<>()
    @     0x7fd65ba22616 std::_Sp_counted_ptr_inplace<>::_M_dispose()
    @     0x55955e44de22 std::_Sp_counted_base<>::_M_release()
    @     0x55955e44dc81 std::__shared_count<>::~__shared_count()
    @     0x7fd65b9df5d9 std::__shared_ptr<>::~__shared_ptr()
    @     0x7fd65ba1ecfc std::__shared_ptr<>::reset()
    @     0x7fd65ba1e367 srf::internal::system::FiberManager::join()
    @     0x7fd65ba1de10 srf::internal::system::FiberManager::~FiberManager()
    @     0x7fd65ba3de9f __gnu_cxx::new_allocator<>::destroy<>()
    @     0x7fd65ba3dd9d std::allocator_traits<>::destroy<>()
    @     0x7fd65ba3d8e0 std::_Sp_counted_ptr_inplace<>::_M_dispose()
    @     0x55955e44de22 std::_Sp_counted_base<>::_M_release()
    @     0x55955e44dc81 std::__shared_count<>::~__shared_count()
    @     0x7fd65ba36521 std::__shared_ptr<>::~__shared_ptr()
    @     0x7fd65ba386d2 std::__shared_ptr<>::reset()
    @     0x7fd65ba36873 srf::internal::system::System::~System()
    @     0x7fd65ba3dbad std::_Sp_counted_ptr<>::_M_dispose()
Aborted (core dumped)
```
from gdb the backtrace looks like:
```
#0  __GI_raise (sig=sig@entry=6) at ../sysdeps/unix/sysv/linux/raise.c:50
#1  0x00007f6d25eaa859 in __GI_abort () at abort.c:79
#2  0x00007f6d26145036 in __gnu_cxx::__verbose_terminate_handler () at ../../../../libstdc++-v3/libsupc++/vterminate.cc:95
#3  0x00007f6d26143524 in __cxxabiv1::__terminate (handler=<optimized out>) at ../../../../libstdc++-v3/libsupc++/eh_terminate.cc:48
#4  0x00007f6d261428c5 in __cxa_call_terminate (ue_header=0x558bdc8c8110) at ../../../../libstdc++-v3/libsupc++/eh_call.cc:54
#5  0x00007f6d26142e6f in __cxxabiv1::__gxx_personality_v0 (version=<optimized out>, actions=<optimized out>, exception_class=5138137972254386944, ue_header=0x558bdc8c8110, 
    context=0x7ffde5ca0a30) at ../../../../libstdc++-v3/libsupc++/eh_personality.cc:688
#6  0x00007f6d2608bd79 in _Unwind_RaiseException_Phase2 (exc=exc@entry=0x558bdc8c8110, context=context@entry=0x7ffde5ca0a30, frames_p=frames_p@entry=0x7ffde5ca0938)
    at ../../../libgcc/unwind.inc:64
#7  0x00007f6d2608c29b in _Unwind_Resume (exc=0x558bdc8c8110) at ../../../libgcc/unwind.inc:242
#8  0x00007f6d2630a238 in google::LogMessage::Flush() [clone .cold] () from /opt/conda/envs/srf/lib/libglog.so.1
#9  0x00007f6d26313eaa in google::LogMessageFatal::~LogMessageFatal() () from /opt/conda/envs/srf/lib/libglog.so.1
#10 0x0000558bdb38e736 in srf::runnable::LaunchControl::prepare_launcher<TestFiberRunnable> (this=0x7f6cf8041c30, options=..., runnable=...)
    at /work/include/srf/runnable/launch_control.hpp:203
#11 0x0000558bdb38cdf3 in TestRunnable_FiberRunnableMisMatch_Test::TestBody (this=0x558bdc8ad010) at /work/src/tests/test_runnable.cpp:252
#12 0x00007f6d2750798e in void testing::internal::HandleExceptionsInMethodIfSupported<testing::Test, void>(testing::Test*, void (testing::Test::*)(), char const*) ()
   from /opt/conda/envs/srf/lib/libgtest.so
#13 0x00007f6d27507be1 in testing::Test::Run() () from /opt/conda/envs/srf/lib/libgtest.so
#14 0x00007f6d27507f0f in testing::TestInfo::Run() () from /opt/conda/envs/srf/lib/libgtest.so
#15 0x00007f6d27508036 in testing::TestSuite::Run() () from /opt/conda/envs/srf/lib/libgtest.so
#16 0x00007f6d275085ec in testing::internal::UnitTestImpl::RunAllTests() () from /opt/conda/envs/srf/lib/libgtest.so
#17 0x00007f6d27508859 in testing::UnitTest::Run() () from /opt/conda/envs/srf/lib/libgtest.so
#18 0x0000558bdb25d16b in RUN_ALL_TESTS () at /opt/conda/envs/srf/include/gtest/gtest.h:2473
#19 0x0000558bdb25d277 in main (argc=1, argv=0x7ffde5ca11a8) at /work/src/tests/test_main.cpp:38
```

**Steps/Code to reproduce bug**
`./src/tests/test_srf_private.x`

**Expected behavior**
Passing

**Environment overview (please complete the following information)**
 - Environment location: Docker
 - Method of SRF install: from source
   - If method of install is [Docker], provide `docker pull` & `docker run` commands used
```bash
DOCKER_BUILDKIT=1 docker build --label dg_dev --network=host --target development -t srf:david_main .
docker run --rm -ti --gpus all --cap-add=sys_nice -v $PWD:/work --label dg_dev -v ${SSH_AUTH_SOCK}:/ssh-agent:ro -e SSH_AUTH_SOCK=/ssh-agent --net host srf:david_main bash
cmake -B build -GNinja -DCMAKE_BUILD_TYPE=Debug -DCMAKE_MESSAGE_CONTEXT_SHOW=ON -DSRF_USE_IWYU=OFF -DSRF_BUILD_DOCS=ON -DSRF_BUILD_EXAMPLES=ON -DSRF_USE_CONDA=ON -DSRF_BUILD_BENCHMARKS=ON .
cmake --build build
pip install build/python
cd build
./src/tests/test_srf_private.x
```

### LINKED ISSUE #36: [BUG] Race Condition in System and Resource Shutdown
**Describe the bug**

Likely a race condition in the shutdown of System and Resources. 

This probably creeped by relaxing the use of `shared_ptr` in lieu of reference.  

Relates to #35 

**Steps/Code to reproduce bug**

Hard to hit has it's a race condition, but if #35 can be consistently reproduced on the given CI system, then it will be easier to iterate.

**Expected behavior**

Clean shutdown.

**Environment overview (please complete the following information)**
**Environment details**
See #35 

Blocks #35

### PR DESCRIPTION
**Pull request by ryanolson**
Fixes #35 
Fixes #36 

The primary issue that was causing #36 is shared_ptr creep. Usually the deferred deletion / pseudo-garbage collection enabled by shared_ptrs is useful, but in this case, the resources that own the worker threads are held in objects that are managed by shared_ptrs. In some cases, the worker threads are the last owners of these shared ptrs and thus exhibit a deadlock where a thread ultimately tries to join itself.

This PR removes all shared_ptrs from object/resources that could be potentially captured by worker threads and replaces them with references.

The code is structured so that all resources and worker threads are owned by the `Executor` -> `resources::Manager` and that all threads must be joined before `resources::Manager` is destroyed. 

This ensures that the resources outlive the pipeline execution provided that the Executor stays in scope. If it goes out of scope or is destroyed it should issue a kill to the pipeline, but still block until all threads are joined.

All executor resources should be created and managed by `resources::Manager` as a single point for life cycle management.

### CODE REVIEW

File: src/internal/resources/manager.cpp [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+  SystemProvider(*resources),
+  m_system(std::move(resources))
+{
+    // for each host partition, construct the runnable resources
+    for (std::size_t i = 0; i < this->system().partitions().host_partitions().size(); ++i)
+    {
+        m_runnable.emplace_back(*m_system, i);
+    }
+
+    // if network, then for each flattened partition
```
Comment (mdemoret-nv) [inline src/internal/resources/manager.cpp:43]: Missing something here? Comment seems like something should follow this.

File: src/internal/runnable/engine_factory.cpp [hunk 2]
```diff
@@ -61,7 +63,7 @@ class FiberEngineFactory : public ::srf::runnable::EngineFactory
     }
 
   private:
-    virtual std::vector<std::shared_ptr<core::FiberTaskQueue>> get_next_n_queues(std::size_t count) = 0;
+    virtual std::vector<std::reference_wrapper<core::FiberTaskQueue>> get_next_n_queues(std::size_t count) = 0;
```
Comment (mdemoret-nv) [inline src/internal/runnable/engine_factory.cpp:66]: I understand that you dont want to use a `shared_ptr` here to avoid a deadlock or circular references, but would a `weak_ptr` work here as well?

Whatever caused the circular reference due to lambda capture is capturing it because it needs it during destruction. Using a reference here cleans up ownership but doesnt guarantee that the object stays alive while its being used. A `weak_ptr` wouldnt influence the lifetime of the object similar to a reference, but you could make the guarantee that the object stayed alive while you were using it.

File: src/internal/runnable/resources.cpp [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+#include "internal/runnable/engine_factory.hpp"
+#include "internal/runnable/engines.hpp"
+#include "internal/system/engine_factory_cpu_sets.hpp"
+#include "internal/system/host_partition.hpp"
+#include "internal/system/partitions.hpp"
+#include "internal/system/system.hpp"
+
+#include "srf/core/bitmap.hpp"
+#include "srf/runnable/types.hpp"
+#include "srf/types.hpp"
```
Comment (mdemoret-nv) [inline src/internal/runnable/resources.cpp:31]: These should all use `<>`

File: src/internal/runnable/resources.cpp [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "internal/runnable/resources.hpp"
+
+#include <srf/runnable/launch_control_config.hpp>
```
Comment (mdemoret-nv) [inline src/internal/runnable/resources.cpp:20]: This should be down with the other `#include <srf/...>`

File: src/internal/runnable/resources.hpp [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+#pragma once
+
+#include "internal/system/fiber_task_queue.hpp"
+#include "internal/system/partition.hpp"
+#include "internal/system/resources.hpp"
+#include "internal/system/system_provider.hpp"
+
+#include "srf/core/task_queue.hpp"
+#include "srf/pipeline/resources.hpp"
+#include "srf/runnable/launch_control.hpp"
```
Comment (mdemoret-nv) [inline src/internal/runnable/resources.hpp:27]: These should use `<>`

File: src/internal/segment/builder.cpp [hunk 6]
```diff
@@ -114,7 +114,7 @@ void Builder::add_object(const std::string& name, std::shared_ptr<::srf::segment
     m_objects[name] = std::move(object);
 }
 
-void Builder::add_runnable(const std::string& name, std::shared_ptr<runnable::Launchable> runnable)
+void Builder::add_runnable(const std::string& name, std::shared_ptr<srf::runnable::Launchable> runnable)
```
Comment (mdemoret-nv) [inline src/internal/segment/builder.cpp:117]: Why does this need to be fully qualified?

File: src/internal/system/fiber_pool.hpp [hunk 7]
```diff
@@ -19,9 +19,12 @@
 
 #include "internal/system/fiber_task_queue.hpp"
 
+#include <functional>
+
```
Comment (mdemoret-nv) [inline src/internal/system/fiber_pool.hpp:23]: Should be after `srf`

File: src/internal/system/fiber_task_queue.cpp [hunk 8]
```diff
@@ -18,7 +18,8 @@
 #include "internal/system/fiber_task_queue.hpp"
 
 #include "internal/system/fiber_priority_scheduler.hpp"
-#include "internal/system/system.hpp"
+#include "internal/system/resources.hpp"
+#include "srf/core/bitmap.hpp"
 #include "srf/core/fiber_meta_data.hpp"
 #include "srf/core/task_queue.hpp"
 #include "srf/types.hpp"
```
Comment (mdemoret-nv) [inline src/internal/system/fiber_task_queue.cpp:25]: These should be separated from the other includes and use `<>`

File: src/internal/system/fiber_task_queue.hpp [hunk 9]
```diff
@@ -17,26 +17,30 @@
 
 #pragma once
 
-#include <srf/core/task_queue.hpp>
+#include "internal/system/thread.hpp"
+#include "srf/utils/macros.hpp"
```
Comment (mdemoret-nv) [inline src/internal/system/fiber_task_queue.hpp:22]:
```suggestion

#include <srf/utils/macros.hpp>
```

File: src/internal/system/iresources.cpp [hunk 10]
```diff
@@ -0,0 +1,46 @@
...
+IResources::~IResources() = default;
+
+void IResources::add_thread_initializer(std::function<void()> initializer_fn)
+{
+    m_impl->register_thread_local_initializer(m_impl->system().topology().cpu_set(), std::move(initializer_fn));
+}
+
+void IResources::add_thread_finalizer(std::function<void()> finalizer_fn)
+{
+    m_impl->register_thread_local_finalizer(m_impl->system().topology().cpu_set(), std::move(finalizer_fn));
```
Comment (mdemoret-nv) [inline src/internal/system/iresources.cpp:43]: If this is a PIMPL, then this shouldnt do any logic here. It should just forward to the `m_impl`.

File: src/internal/system/partition.cpp [hunk 11]
```diff
@@ -0,0 +1,59 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "internal/system/partition.hpp"
+
+#include "srf/exceptions/runtime_error.hpp"
```
Comment (mdemoret-nv) [inline src/internal/system/partition.cpp:20]:
```suggestion
#include <srf/exceptions/runtime_error.hpp>
```

File: src/internal/system/partitions.cpp [hunk 12]
```diff
@@ -17,9 +17,12 @@
...
 #include "internal/system/partitions.hpp"
 
+#include "internal/system/partition.hpp"
 #include "internal/system/system.hpp"
 #include "internal/system/topology.hpp"
 #include "internal/utils/shared_resource_bit_map.hpp"
+
+#include "srf/core/bitmap.hpp"
 #include "srf/options/placement.hpp"
 #include "srf/utils/bytes_to_string.hpp"
```
Comment (mdemoret-nv) [inline src/internal/system/partitions.cpp:27]: These should use `<>`

File: src/internal/system/partitions.hpp [hunk 13]
```diff
@@ -31,23 +32,27 @@ namespace srf::internal::system {
 class System;
 class Topology;  // IWYU pragma: keep
 
+// TODO(unknown): flatten partition here
+
```
Comment (mdemoret-nv) [inline src/internal/system/partitions.hpp:36]: This should be `TODO(ryan)`

File: src/internal/system/partitions.hpp [hunk 14]
```diff
@@ -31,23 +32,27 @@ namespace srf::internal::system {
...
 class Partitions
 {
   public:
     Partitions(const Topology& topology, const Options& options);
     Partitions(const System& system);
 
+    const std::vector<Partition>& flattened() const;
```
Comment (mdemoret-nv) [inline src/internal/system/partitions.hpp:43]: Why `flattened()`? Why not just `partitions()` or `all_partitions()`?

File: src/internal/system/system.cpp [hunk 15]
```diff
@@ -17,27 +17,18 @@
 
 #include "internal/system/system.hpp"
 
-#include "internal/system/fiber_task_queue.hpp"
 #include "internal/system/partitions.hpp"
-#include "public/utils/thread_utils.hpp"
 
+#include "srf/core/bitmap.hpp"
```
Comment (mdemoret-nv) [inline src/internal/system/system.cpp:22]: These should use `<>`

File: src/internal/system/system_provider.cpp [hunk 16]
```diff
@@ -0,0 +1,36 @@
...
+
+SystemProvider::SystemProvider(std::shared_ptr<const System> system) : m_system(std::move(system))
+{
+    CHECK(m_system);
+}
+const System& srf::internal::system::SystemProvider::system() const
+{
+    DCHECK(m_system);
+    return *m_system;
+}
```
Comment (mdemoret-nv) [inline src/internal/system/system_provider.cpp:34]: Why is this needed?

**Review by mdemoret-nv — CHANGES_REQUESTED** [review]
This seems like a lot of changes to fix a deadlock on resource shutdown and brings up some concerns about how intertwined the design is.

Also, I dont see a new test checking that the deadlock has been resolved.

File: include/srf/internal/executor/iexecutor.hpp [hunk 17]
```diff
@@ -24,19 +24,29 @@
...
+ *
+ * Over time, new customization points will be added between ISystemResources and IExecutor.
+ */
 class IExecutor
 {
   public:
     IExecutor();
     IExecutor(std::shared_ptr<Options>);
-    IExecutor(std::unique_ptr<system::ISystem>);
+    IExecutor(std::unique_ptr<system::IResources>);
```
[the remark this answers is not in the document]
Comment (ryanolson) [reply include/srf/internal/executor/iexecutor.hpp:49]: Yes, `IResources` replaces `ISystem` as the object which adds the python thread initializers and finalizers

File: include/srf/internal/system/iresources.hpp [hunk 18]
```diff
@@ -0,0 +1,43 @@
...
+
+class IResources
+{
+  public:
+    IResources(std::shared_ptr<ISystem> system);
+    virtual ~IResources() = 0;
+
+  protected:
+    void add_thread_initializer(std::function<void()> initializer_fn);
+    void add_thread_finalizer(std::function<void()> finalizer_fn);
```
[the remark this answers is not in the document]
Comment (ryanolson) [reply include/srf/internal/system/iresources.hpp:36]: you can register as many as you want

File: src/internal/data_plane/instance.cpp [hunk 19]
```diff
@@ -29,69 +30,74 @@
...
+            // if the PartitionResource has a GPU, ensure the CUDA context on the main thread is active
+            // before the ucx context is constructed
+            if (m_resources->device())
+            {
+                auto device = m_resources->device()->get();
+                device.activate();
+
+                void* addr = nullptr;
+                SRF_CHECK_CUDA(cudaMalloc(&addr, 1024));
+                SRF_CHECK_CUDA(cudaFree(addr));
```
[the remark this answers is not in the document]
Comment (ryanolson) [reply src/internal/data_plane/instance.cpp:66]: This is a stronger guarantee. And since it's not in the hot path, paying the cost of N small device mallocs and frees is insignificant.

File: src/internal/system/system_provider.cpp (hunk 16 above)
Comment (ryanolson) [reply src/internal/system/system_provider.cpp:34]: To what are you referring?

The `CHECK`?  The `DCHECK`? or the `SystemProvider` object?

File: CMakeLists.txt [hunk 20]
```diff
@@ -197,16 +197,17 @@ add_library(libsrf
...
-  src/internal/resources/device_resources.cpp
-  src/internal/resources/host_resources.cpp
-  src/internal/resources/partition_resources.cpp
-  src/internal/resources/resource_partitions.cpp
-  src/internal/resources/system_resources.cpp
+  src/internal/resources/manager.cpp
+# src/internal/resources/host_resources.cpp
+# src/internal/resources/partition_resources.cpp
+# src/internal/resources/resource_partitions.cpp
+# src/internal/resources/system_resources.cpp
```
[the remark this answers is not in the document]
Comment (ryanolson) [reply CMakeLists.txt:204]: If you don't mind, these will be addressed in one of the first PRs for 22.08.  Re-establishing the architect service requires touching all these files, which is in progress on my `network_resources` branch.

File: src/internal/resources/manager.cpp (hunk 1 above)
Comment (ryanolson) [reply src/internal/resources/manager.cpp:43]: The next edition is adding network  resources. There is a development branch in my repo, but I haven't made a PR yet to avoid unnecessary CI runs.

File: include/srf/internal/system/iresources.hpp (hunk 18 above)
[the remark this answers is not in the document]
Comment (mdemoret-nv) [reply include/srf/internal/system/iresources.hpp:36]: Do they overwrite the previous one or are they executed in any sort of order?

File: include/srf/internal/executor/iexecutor.hpp (hunk 17 above)
[the remark this answers is not in the document]
Comment (mdemoret-nv) [reply include/srf/internal/executor/iexecutor.hpp:49]: Ok

File: CMakeLists.txt (hunk 20 above)
[the remark this answers is not in the document]
Comment (mdemoret-nv) [reply CMakeLists.txt:204]: Can you add a comment in the CMakeLists.txt file indicating why they are commented out with a link to an issue? I really hate having commented out source files in the CMakeLists.txt.

File: src/internal/runnable/engine_factory.cpp (hunk 2 above)
Comment (ryanolson) [reply src/internal/runnable/engine_factory.cpp:66]: I've restructured the code to make strong guarantees on the order creation and destruction for all resources. 

This will allow for more structured concurrency as we move toward coroutines. 

The only requirement is the Executor object is [uniquely] owned by a thread that is not constructed as part of the SRF runtime. 

Remember, you can't use a weak_ptr directly. In order to use a weak_ptr, you need to construct a shared_ptr via the lock method.

File: src/internal/segment/builder.cpp (hunk 6 above)
Comment (ryanolson) [reply src/internal/segment/builder.cpp:117]: If you are in srf::internal::xx and there is srf::xx, the you need to fully quality objects from the public NS.

File: src/internal/system/partitions.hpp (hunk 14 above)
Comment (ryanolson) [reply src/internal/system/partitions.hpp:43]: There is a comment explaining this on one of the files. 

The host and device partition are hierarchical where there is a possibility, depending on option where more than one device shared the same host partition, so those host resources are shared. 

However, each device defines a partition since it will have its own unique network and device resources. 

Just we flatten the partitions such that it's a flat list where the first N partitions correspond to N cuda devices. If there are more than N partition, those partitions do not have devices attached. 

The flattened partitions provide the view of what's available to a given GPU. 

The plan is to exposes the set of flattened partitions and their resources to the builder and the runtime respectively.

File: include/srf/internal/system/iresources.hpp (hunk 18 above)
[the remark this answers is not in the document]
Comment (ryanolson) [reply include/srf/internal/system/iresources.hpp:36]: Stored using a multimap, order is not guaranteed, nor should it be expected. 

These lambda should be independent of each other.

File: src/internal/system/system_provider.cpp (hunk 16 above)
Comment (mdemoret-nv) [reply src/internal/system/system_provider.cpp:34]: Just the class in general. I think I figured it out.

File: src/internal/system/system.cpp (hunk 15 above)
Comment (mdemoret-nv) [reply src/internal/system/system.cpp:22]: Will be fixed in PR #47

File: src/internal/system/partitions.hpp (hunk 14 above)
Comment (mdemoret-nv) [reply src/internal/system/partitions.hpp:43]: Can we put this in a comment on that property?

File: src/internal/system/partitions.cpp (hunk 12 above)
Comment (mdemoret-nv) [reply src/internal/system/partitions.cpp:27]: Will be fixed in PR #47

File: include/srf/internal/system/iresources.hpp (hunk 18 above)
[the remark this answers is not in the document]
Comment (mdemoret-nv) [reply include/srf/internal/system/iresources.hpp:36]: Ok, `add_` works then.

File: src/internal/runnable/engine_factory.cpp (hunk 2 above)
Comment (mdemoret-nv) [reply src/internal/runnable/engine_factory.cpp:66]: > Remember, you can't use a weak_ptr directly. In order to use a weak_ptr, you need to construct a shared_ptr via the lock method.

Thats exactly what I am saying. If you have some code like this:
```cpp
// Get a single queue. next_queue will be of type std::reference_wrapper<core::FiberTaskQueue>
auto next_queue = this->get_next_n_queues(1)[0];

// Do some action with the queue.
next_queue.affinity();

// Sleep the fiber, when it wakes up next_queue could be destroyed
boost::this_fiber::sleep(1000);

// Try to perform some action on the reference again
next_queue.affinity(); // Crash
```
There is no guarantee that while you are holding onto `next_queue` it wont get destructed. It would have before since it was a `shared_ptr`.

Now if you used weak_ptr:
```cpp
// Get a single queue. next_queue will be of type std::weak_ptr<core::FiberTaskQueue>
auto next_queue = this->get_next_n_queues(1)[0];

// Need to convert my weak_ptr to a shared_ptr
std::shared_ptr<core::FiberTaskQueue> next_queue_ptr = next_queue.lock();

// Do some action with the queue.
next_queue_ptr->affinity();

// Sleep the fiber, when it wakes up next_queue_ptr is guaranteed to be alive
boost::this_fiber::sleep(1000);

// Try to perform some action on the reference again
next_queue_ptr->affinity();

// next_queue_ptr goes out of scope and can be destroyed
```

File: src/internal/segment/builder.cpp (hunk 6 above)
Comment (mdemoret-nv) [reply src/internal/segment/builder.cpp:117]: We need to fix up these duplicate names

File: src/internal/system/fiber_task_queue.cpp (hunk 8 above)
Comment (mdemoret-nv) [reply src/internal/system/fiber_task_queue.cpp:25]: Will be fixed in PR #47

File: src/internal/system/fiber_pool.hpp (hunk 7 above)
Comment (mdemoret-nv) [reply src/internal/system/fiber_pool.hpp:23]: Will be fixed in PR #47

File: src/internal/runnable/resources.hpp (hunk 5 above)
Comment (mdemoret-nv) [reply src/internal/runnable/resources.hpp:27]: Will be fixed in PR #47

File: src/internal/runnable/resources.cpp (hunk 4 above)
Comment (mdemoret-nv) [reply src/internal/runnable/resources.cpp:20]: Will be fixed in PR #47

File: src/internal/runnable/resources.cpp (hunk 3 above)
Comment (mdemoret-nv) [reply src/internal/runnable/resources.cpp:31]: Will be fixed in PR #47

File: src/internal/system/partitions.hpp (hunk 14 above)
Comment (ryanolson) [reply src/internal/system/partitions.hpp:43]: Done

File: src/internal/system/iresources.cpp (hunk 10 above)
Comment (ryanolson) [reply src/internal/system/iresources.cpp:43]: the impl methods to add initializers/finalizers are a bit more general, so we are adding one extra argument for the cpu set (the full cpu set) on which we want to apply the initializer/finalizer.

File: src/internal/runnable/engine_factory.cpp (hunk 2 above)
Comment (ryanolson) [reply src/internal/runnable/engine_factory.cpp:66]: There are strong guarantees that the FiberTaskQueues outlive all pipeline/segment/node/runnables.

Holding a shared pointer to a FiberTaskQueue which owns `system::Thread` on the thread/fiber is the culprit of the deadlock.

Comment (mdemoret-nv) [reply src/internal/runnable/engine_factory.cpp:66]: Ok. Will just have to trust you on this one.

File: src/internal/system/iresources.cpp (hunk 10 above)
Comment (mdemoret-nv) [reply src/internal/system/iresources.cpp:43]: This is probably fine for now. We need to have a general design for the PIMPL classes anyways.

**Comment by mdemoret-nv** [comment]
@gpucibot merge
